guides/hack/21-XHP/01-introduction.md

```diff
@@ -1,8 +1,71 @@
-**Topics covered in this section**
-* [An overview](some-basics.md)
-* [Extending XHP classes](extending.md)
-* [Getting started](basic-usage.md)
-* [Guidelines for use](guidelines.md)
-* [Migrating to XHP](migration.md)
-* [XHP interfaces](interfaces.md)
-* [XHP object methods](methods.md)
+XHP provides a native XML-like representation of output (which is usually HTML). This allows UI code to be typechecked, and automatically
+avoids several common issues such as cross-site scripting (XSS) and double-escaping. It also applies other validation rules, e.g., `<head>`
+must contain `<title>`.
+
+Using traditional interpolation, a simple page could look like this:
+
+```Hack
+$user_name = 'Fred';
+echo "<tt>Hello <strong>$user_name</strong></tt>";
+```
+
+However, with XHP, it looks like this:
+
+```
+$user_name = 'Fred';
+echo <tt>Hello <strong>{$user_name}</strong></tt>;
```

Comment:
`echo` won't work here, you need to call `->toStringAsync()`

Reply:
we should generally try to remove all of these inline snippets and have them in external files so they're typechecked

Reply:
Good catch, I'll fix it.

> we should generally try to remove all of these inline snippets and have them in external files so they're typechecked

Agreed but the way it currently works, it would be very annoying for all these short examples. Maybe we could add a build step that extracts examples from the text (I think you suggested something like that before)?

Reply:
Yeah, I think the best way to do this is a github action that:

1. extracts
2. generates output
3. automatically creates a pull request against the pull request origin branch

... I'm not sure where the source repo would be for 3 though - don't really want it to be this repo.

Reply:
Why not do it in bin/build.php? Ideally I'd want my examples tested before I commit them.

Reply:
To make things easier for people who approach it as "I'm just editing markdown, no need for the full dev environment" - i.e. they use the github web editor, or just edit the markdown files without running the test.

![Unknown](https://user-images.githubusercontent.com/360927/89562817-691e0a80-d7cf-11ea-9388-4ffee5e7ef1c.jpeg)

Do it in build.php, but make the CI check that build.php has no committable changes, and submit a PR-against-the-PR if not?

if done in build.php, it should prompt (unless some env var/flag is provided), rather than silently modifying source files

Reply:
> if done in build.php, it should prompt (unless some env var/flag is provided), rather than silently modifying source files

I think I'd rather keep the auto-generated output out of source files completely (no part of the source .md file should be generated), we should keep it in .out/.expect or similar files that get pulled into the final HTML during build.

- build.php would generate any out/expect files that are missing, leave existing ones alone.
- hacktest would verify that out/expect file exists for every example and that the output matches.